Add CHANGELOG entry for SE-0235.

jshier opened the pull request:
This PR simply adds a `CHANGELOG` entry for the addition of `Result` in SE-0235.

I was unsure of how much detail or if mention of the `Error` self-conformance or type shadowing changes made in relation to adding `Result` were needed, so I left them out for now.

Azoy commented:
cc: @natecook1000

xwu commented on CHANGELOG.md:
````diff
@@ -24,6 +24,19 @@ CHANGELOG
+  The standard library now contains a `Result<Success, Failure: Error>` type to manually model potentially error-producing code.
+  
+  ```swift
+  enum Result<Success, Failure: Error>
````
```suggestion
  enum Result<Success, Failure: Error> {
```

xwu commented on CHANGELOG.md:
````diff
@@ -24,6 +24,19 @@ CHANGELOG
+  }
+  ```
+  
+  This type can be used in contexts where throwing or catching errors may be undesired or impossible while still using `Error`-conforming types, such as asynchronous methods or methods where Swift's automatic error propagation isn't ergonomic.
````
I think this text could more closely reflect the guidance on the Forums surrounding when to use `Result` versus throwing.

jshier replied:
Could outline the points you'd like to see?

xwu replied:
"Undesired or impossible" is pretty vague. Some well received answers on that point--

@jckarter wrote:
> I think it's reasonable to see Result and throw as serving different needs within the general realm of error handling and propagation. As I'd mentioned in the previous thread, throw was intentionally designed with an eye toward loosely-coupled, handled-at-a-distance failures, which is why it features automatic propagation (to minimize the overhead of carrying the error to the handler) and no typing (because types would not really help you handle all uncountably many forms of network/file/hardware/etc. error that could accrete through layers of a system). A lot of the people reaching for Result or wishing for typed throws are dealing with more immediate error conditions for which implicit propagation and try/catch ceremony are overkill, and being able to be more specific about the result type is valuable to contain the set of error conditions.

You wrote:
> Result does indeed provide something new to the language: manually propagated error handing. While use of this case is not going to be as popular as the automatic try/throw, it has uses that that construct can't handle.

jshier commented:
Updated the `Result` text and added a note about `Error` self-conformance.

Two questions: is there a way I'm supposed to refer to Swift's current error handling? I see `try` / `catch` and `throws` used separately and it's not clear which I should use. Also, I thought my phrasing for the `Error` self conformance was awkward. Is there a better way to state it with more detail than just "making it more useful in generic contexts"?

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
 
+* [SE-0235][]:
+
+  The standard library now contains a `Result<Success, Failure: Error>` type to 
```
```suggestion
  The standard library now contains a `Result` type for manually propagating errors.
```

(It is unnecessary to repeat the generic parameters here when it is presented one line later.)

It is unclear what "manually model" means; all modeling is manual, but what's different here is how errors are propagated. "Error-producing code" sounds to me like code that's ill-formed, producing compiler errors.

xwu commented on CHANGELOG.md:
````diff
@@ -24,6 +24,28 @@ CHANGELOG
+  }
+  ```
+  
+  This type serves a separate but complementary role to Swift's existing and 
````
```suggestion
  This type serves a complementary role to that of `throws`.
```

"Complementary" implies "separate"; and you do not want to echo the phrase "separate but equal."  
"Swift" is unnecessary because it is implied.  
"Existing" is unnecessary because one would hardly bring up non-existent features for comparison.

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
+* [SE-0235][]:
+
+  The standard library now contains a `Result<Success, Failure: Error>` type to 
+  manually model potentially error-producing code.
```
```suggestion
```

xwu commented on CHANGELOG.md:
````diff
@@ -24,6 +24,28 @@ CHANGELOG
+  ```
+  
+  This type serves a separate but complementary role to Swift's existing and 
+  automatic error handling using `try` / `catch`. In addition to it's usefulness 
````
```suggestion
```

Describing how they differ can be concentrated into the following sentence; there does not need to be a phrase to introduce well known language keywords.

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
+  
+  This type serves a separate but complementary role to Swift's existing and 
+  automatic error handling using `try` / `catch`. In addition to it's usefulness 
+  in asynchronous code, `Result` can be used in situations where the formality of
```
```suggestion
  `Result` can be used where implicit error propagation and the ceremony of
```

"In situations where" is redundant.  
Specific examples can come afterwards instead of sandwiching the explanation of the examples.  
"Formality" has ambiguous connotations.

jshier replied:
I swapped "automatic" for "implicit" in your suggested change, as I don't think the use of explicit keywords is very implicit, and I've seen "automatic" more commonly used to describe Swift's error handling, especially in the error manifesto.

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
+  This type serves a separate but complementary role to Swift's existing and 
+  automatic error handling using `try` / `catch`. In addition to it's usefulness 
+  in asynchronous code, `Result` can be used in situations where the formality of
+  `try` / `catch` is undesirable. Additionally, it can be used in situations that 
```
```suggestion
  `try` / `catch` are undesirable, such as in asynchronous code or
```

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
+  automatic error handling using `try` / `catch`. In addition to it's usefulness 
+  in asynchronous code, `Result` can be used in situations where the formality of
+  `try` / `catch` is undesirable. Additionally, it can be used in situations that 
+  aren't easily support by `try` / `catch`, such as  accumulating the results of 
```
```suggestion
  in accumulating the results of 
```

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
+  in asynchronous code, `Result` can be used in situations where the formality of
+  `try` / `catch` is undesirable. Additionally, it can be used in situations that 
+  aren't easily support by `try` / `catch`, such as  accumulating the results of 
+  several sequential error-producing operations.
```
```suggestion
  successive error-producing operations.
```

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
+  aren't easily support by `try` / `catch`, such as  accumulating the results of 
+  several sequential error-producing operations.
+  
+* `Error` now conforms to itself. This allows the use of `Error` values in generic
```
```suggestion
* `Error` now conforms to itself. This allows for the use of `Error` itself as the
```

xwu commented on CHANGELOG.md:
```diff
@@ -24,6 +24,28 @@ CHANGELOG
+  several sequential error-producing operations.
+  
+* `Error` now conforms to itself. This allows the use of `Error` values in generic
+  contexts without a separate, concrete conforming type.
```
```suggestion
  argument for a generic parameter constrained to `Error`.
```

natecook1000 commented on CHANGELOG.md:
```diff
@@ -24,6 +24,25 @@ CHANGELOG
+  This type serves a complementary role to that of `throws`. `Result` can be used 
+  in situations where automatic error propagation or the ceremony of `try` / `catch` 
+  are undesirable, such as in asynchronous code or in accumulating the results of 
+  successive error-producing operations.
```
How about this? I'd like to avoid the passive voice and calling `try`-`catch` a "ceremony":

> This type serves a complementary role to that of throwing functions and initializers. Use `Result` in situations where automatic error propagation or `try`-`catch` blocks are undesirable, such as in asynchronous code or when accumulating the results of successive error-producing operations.

jshier committed:
Add CHANGELOG entry for Result.

jshier commented:
@natecook1000 Updated.

natecook1000 commented:
@swift-ci Please smoke test and merge

benrimmington commented on CHANGELOG.md:
```diff
@@ -7406,6 +7425,7 @@ Swift 1.0
 [SE-0227]: <https://github.com/apple/swift-evolution/blob/master/proposals/0227-identity-keypath.md>
 [SE-0228]: <https://github.com/apple/swift-evolution/blob/master/proposals/0228-fix-expressiblebystringinterpolation.md>
 [SE-0230]: <https://github.com/apple/swift-evolution/blob/master/proposals/0230-flatten-optional-try.md>
+[SE-O235]: <https://github.com/apple/swift-evolution/blob/master/proposals/0235-add-result.md>
```
Typo (LATIN CAPITAL LETTER O should be DIGIT ZERO):

```diff
-[SE-O235]: <https://github.com/apple/swift-evolution/blob/master/proposals/0235-add-result.md>
+[SE-0235]: <https://github.com/apple/swift-evolution/blob/master/proposals/0235-add-result.md>
```

natecook1000 replied:
Oy! Fixed in #22879.
